### PR TITLE
Integrate Turan Oracle into Casper logs

### PR DESCRIPTION
**Pull request by KentShikama**
## Overview
Provide a brief description of what this PR does, and why it's needed.

Relative to #879 

Runs turan oracle over the forkchoice everytime a new block is added (for now). The log output looks like follows:

```
15:29:24.151 [grpc-default-executor-0] INFO  logger - CASPER: Beginning send of Block #22 (47ee0a536e...) -- Sender ID a4b7a459ce... -- M Parent Hash d674ded414... -- Contents 4540b684eb... to peers...
15:29:24.151 [grpc-default-executor-0] INFO  logger - CASPER: Added 47ee0a536e...
15:29:24.153 [grpc-default-executor-0] INFO  logger - CASPER: New fork-choice tip is block 47ee0a536e....
15:29:24.222 [grpc-default-executor-0] INFO  logger - CASPER: Block 47ee0a536e... has a fault tolerance of -1.0.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block d674ded414... has a fault tolerance of -1.0.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block e49eefcc36... has a fault tolerance of -0.8666667.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block eb26664a20... has a fault tolerance of -0.8666667.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block a2bab7245b... has a fault tolerance of -0.8666667.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 25ebb5b5fa... has a fault tolerance of -0.8666667.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 8f7821c0b7... has a fault tolerance of -0.8666667.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block daf18ed7ac... has a fault tolerance of -0.6.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 4afc56b56d... has a fault tolerance of -0.6.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 5a20b858d7... has a fault tolerance of -0.6.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block def364fe35... has a fault tolerance of -0.6.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block e75d31f271... has a fault tolerance of -0.6.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 44c28730d4... has a fault tolerance of -0.2.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block a7509212bd... has a fault tolerance of -0.2.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 5fda723caf... has a fault tolerance of -0.2.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 75ad5d5fc1... has a fault tolerance of -0.2.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 2c4266bdd0... has a fault tolerance of -0.2.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 66f6a551f3... has a fault tolerance of -0.2.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 37c838bed2... has a fault tolerance of -0.2.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 8a099ce53b... has a fault tolerance of -0.2.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block daaecdeadc... has a fault tolerance of -0.2.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 0e0da5112c... has a fault tolerance of -0.2.
15:29:24.223 [grpc-default-executor-0] INFO  logger - CASPER: Block 7a4998626c... has a fault tolerance of -0.2.
```
The Turan Oracle works best when you have validators with equal weights.

### Does this PR relate to an RChain JIRA issue? 
If applicable, add link to corresponding JIRA issue.

Part of https://rchain.atlassian.net/browse/RHOL-336

### Complete this checklist before you submit the PR
- [X] This PR contains no more than 200 lines of code, excluding test code.
- [X] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [X] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala [hunk 1]
```diff
@@ -30,107 +31,126 @@ import scala.collection
  * and is_safe will reduce to total_weight >= total_weight and evaluate to true.
  */
 trait SafetyOracle[F[_]] {
-  def isSafe(estimate: BlockMessage, faultToleranceThreshold: Float): F[Boolean]
+
+  /**
+    * @param estimate Block to detect safety on
+    * @return normalizedFaultTolerance float between -1 and 1, where -1 means potentially orphaned
```
[the remark this answers is not in the document]
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala]: Yeah it generalizes to any oracle. My internal definition of a safety oracle is a function that takes a blockDAG and an estimate and returns a fault tolerance (or returns the fact that it is safe...but then you have to pass in a FT threshold).

File: casper/src/main/scala/coop/rchain/casper/Casper.scala [hunk 2]
```diff
@@ -88,19 +88,32 @@ sealed abstract class MultiParentCasperInstances {
...
+          tip       = estimates.head
           _ <- Log[F].info(
-                s"CASPER: New fork-choice is block ${PrettyPrinter.buildString(forkchoice.blockHash)}.")
+                s"CASPER: New fork-choice tip is block ${PrettyPrinter.buildString(tip.blockHash)}.")
+          dag <- blockDag
+          // TODO: Replace with getMainChainUntilLastFinalizedBlock
+          forkchoice = getMainChain(dag, tip, IndexedSeq.empty[BlockMessage])
+          forkchoiceSafety <- forkchoice.toList.traverse(block =>
+                               SafetyOracle[F].normalizedFaultTolerance(dag, block))
+          _ <- forkchoice.zip(forkchoiceSafety).toList.traverse {
```
[the remark this answers is not in the document]
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Casper.scala]: Yeah that is the next step. This is the first half of the ticket. I just wanted proof that this is working without having to query each time the various blocks involved, hence the logging.

**Comment by KentShikama** [comment]
@birchmd PTAL

File: casper/src/main/scala/coop/rchain/casper/Casper.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/Casper.scala]: Isn't that what unit tests are for? I would still prefer if we didn't log unnecessary things, but I suppose we can take it out in a future PR when you have the query functionality implemented.

[the remark this answers is not in the document]
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Casper.scala]: Yeah will remove in a future PR indeed.

**Pull request merged**